src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParser.cs

```diff
@@ -9172,156 +9172,166 @@ private async Task WriteXmlFeed(XmlDataFeed feed, TdsParserStateObject stateObj,
             {
-                if (_asyncWrite)
+                if (needBom)
                 {
-                    await writer.WriteAsync((char)TdsEnums.XMLUNICODEBOM).ConfigureAwait(false);
+                    if (_asyncWrite)
+                    {
+                        await writer.WriteAsync((char)TdsEnums.XMLUNICODEBOM).ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        writer.Write((char)TdsEnums.XMLUNICODEBOM);
+                    }
                 }
-                else
+
+                int nWritten = 0;
+                do
                 {
-                    writer.Write((char)TdsEnums.XMLUNICODEBOM);
-                }
-            }
+                    int nRead = 0;
 
-            int nWritten = 0;
-            do
-            {
-                int nRead = 0;
+                    if (_asyncWrite)
+                    {
+                        nRead = await feed._source.ReadBlockAsync(inBuff, 0, constTextBufferSize).ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        nRead = feed._source.ReadBlock(inBuff, 0, constTextBufferSize);
+                    }
 
-                if (_asyncWrite)
-                {
-                    nRead = await feed._source.ReadBlockAsync(inBuff, 0, constTextBufferSize).ConfigureAwait(false);
-                }
-                else
-                {
-                    nRead = feed._source.ReadBlock(inBuff, 0, constTextBufferSize);
-                }
+                    if (nRead == 0)
+                    {
+                        break;
+                    }
 
-                if (nRead == 0)
-                {
-                    break;
-                }
+                    if (_asyncWrite)
+                    {
+                        await writer.WriteAsync(inBuff, 0, nRead).ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        writer.Write(inBuff, 0, nRead);
+                    }
+
+                    nWritten += nRead;
+                } while (!writer.IsComplete);
 
                 if (_asyncWrite)
                 {
-                    await writer.WriteAsync(inBuff, 0, nRead).ConfigureAwait(false);
+                    await writer.FlushAsync().ConfigureAwait(false);
                 }
                 else
                 {
-                    writer.Write(inBuff, 0, nRead);
+                    writer.Flush();
                 }
-
-                nWritten += nRead;
-            } while (!writer.IsComplete);
-
-            if (_asyncWrite)
-            {
-                await writer.FlushAsync().ConfigureAwait(false);
-            }
-            else
-            {
-                writer.Flush();
             }
+
+            Array.Clear(inBuff, 0, constTextBufferSize);
```

Comment:
Any particular reason for the Clear? Generally we don't bother clearing before returning to the pool. (Although maybe we should test in a mode where we garbelize after return https://github.com/dotnet/coreclr/issues/9885)

Reply:
It doesn't provide any security benefit that I see.

Reply:
> Any particular reason for the Clear? Generally we don't bother clearing before returning to the pool.

We need to get consistent on this in general.

@danmosemsft is correct that this has been our general stance, and we don't clear in methods like `Stream.CopyToAsync`.  We do clear in code known to be sensitive, e.g. in the crypto stack.  However, depending on who the dev is and what their philosophy is, there's a bunch of places recently where clears have been added in situations where they in theory fit into the former case, e.g. in all of the new json-related code (cc: @bartonjs, @ahsonkhan, @steveharter).

We should all get in agreement and come up with very clear rules, then audit all of our usage and make sure we're in compliance with those rules.

Reply:
Personally, I would prefer if the answer was "always use clearArray: true, except for in specific cases where you want to swear in blood that there's next-to-no benefit to seeing this data" (e.g. JsonDocument's database block, since the most it conveys is the length of a given value -- though I'd be just as happy clearing that one, too).  Next happiest is "always clear what you wrote" (which is the part that I've been talked down to :smile:).

Essentially, as a security person, I want anything that touches user data to clear it before putting it back into the pool; because one of the best things about .NETs memory model is it's very very difficult to get at "discarded" user data.

The litmus test is "assume some web application has a bug where they share the entire contents of a rented array instead of just the portion they meant to... would it be bad for anyone if this data was in that array?".  Unless the answer is "I am certain the answer is no" then treat it as if the answer is yes.

Reply:
Assuming that the user might want to pass sensitive data into the sql server at some point i thought it best to err on the side of caution and clear the part of the buffer i've used which is why i clear what i know i've rented and not the entire buffer. 

Things like string formatting and other internal uses i wouldn't feel the need to clear the array but for this purpose where i don't know what the data contents means and i can't be certain about the lifetime of the array because the purpose of the rental is to be longer lived it seemed appropriate.

Reply:
Fair enough. @stephentoub what do you think of this kind of defense in depth? If I can pull arrays from the pool I presumably already own you. On the other hand, I have heard it argued that it makes it slightly less likely to appear in a crash dump.
I think we have not done this elsewhere. Of course it is harmless if it is not measurable.

Reply:
Uh, ignore my question above @stephentoub . I somehow didn't see your earlier response. If this buffer might contain something sensitive at this point, it seems reasonable to clear it as you have done @Wraith2  - and we should do an all up pass at some point to be more consistent.

Reply:
I have wondered If might be a good idea to have a disposable wrapper for this pattern where clearing of the requested length is the default behaviour.

Reply:
There are at least 2 scenarios where clearing the array yourself is worse than just using the `clearArray` behavior of the `ArrayPool`

1)  When the buffer isn't actually going to be accepted back into the pool.  Only the `ArrayPool` knows whether it will actually recycle the buffer, and it delays clearing until it knows the buffer will be returned to the pool.
2)  When the buffer is an odd size.  If you request a buffer of 63 bytes and are allocated 64, it's obviously cheaper to let the `ArrayPool` clear the entire 64 bytes than it is to clear the 63 yourself.

I'd say even if you're sure you need the array cleared for safety, there should be demonstrable benefits to taking that responsibility away from the `ArrayPool` before it's done

Reply:
Perhaps ArrayPool ought to have an overload which takes the size to clear (as a hint)

Reply:
> When the buffer isn't actually going to be accepted back into the pool.

If the buffer has previously been in the pool it may be in a higher generation than 0 and not be cleaned up for an unknown amount of time. If you're using SqlClient there's a pretty good chance you'll get all generations of GC fairly quickly but i'm working on improving that :grin:.

As far as i know the runtime only guarentees that newly allocated memory is zeroed, that doesn't require that collected memory is zeroed so information could hang around in memory for some time. In any case where the buffer contains user information it has the possibility to be sensitive and i think it should be cleared, even with non pooled buffers. 

In cpu performance terms it's a non-issue in SqlClient because it spends 80-90% of it's time waiting on network io.

@danmosemsft it's possible though i think that option came up in the ArrayPool design and was decided against.

Reply:
@danmosemsft @saucecontrol @bartonjs @stephentoub What's the consensus on this comment? Should we keep it as is or rely on ArrayPool<char>.Shared.Return(inBuff, clearArray: true); ?

Reply:
It seems to me that having ArrayPool clear it ought to be perfectly fine and simpler

Reply:
Ok, done.